### PR TITLE
feat: New Module `digital-twins/digital-twins-instance`

### PR DESCRIPTION
**Pull request by ChrisSidebotham**
## Description

Adding new module for Digital Twins instance

## Pipeline Reference

| Pipeline |
| -------- |
|    [avm.res.digital-twins.digital-twins-instance](https://github.com/ChrisSidebotham/bicep-registry-modules/actions/workflows/avm.res.digital-twins.digital-twins-instance.yml)      |

## Type of Change

- [ ] Update to CI Environment or utlities (Non-module effecting changes)
- [x] Azure Verified Module updates:
  - [ ] Bugfix containing backwards compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [ ] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [ ] Update to documentation

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] I have run `Set-AVMModule` locally to generate the supporting module files.
- [ ] My corresponding pipelines / checks run clean and green without any errors or warnings

### CODE REVIEW

File: avm/res/digital-twins/digital-twins-instance/endpoint--event-hub/main.bicep [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+@description('Optional. The EventHub name in the EventHub namespace for identity-based authentication.')
+param entityPath string = ''
+
+@description('Optional. The URL of the EventHub namespace for identity-based authentication. It must include the protocol \'sb://\' (i.e. sb://xyz.servicebus.windows.net).')
+param endpointUri string = ''
+
+@description('Optional. The managed identity definition for this resource.')
+param managedIdentities managedIdentitiesType
+
+var identity = !empty(managedIdentities) ? {
```
Comment (AlexanderSehr) [inline avm/res/digital-twins/digital-twins-instance/endpoint--event-hub/main.bicep:42]: Question: Does this show a warning? If so, I'd suggest to apply the same 'fix' that @krbar implemented in this PR #1238

File: avm/res/digital-twins/digital-twins-instance/endpoint--service-bus/main.bicep [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+param primaryConnectionString string = ''
+
+@description('Optional. SecondaryConnectionString of the endpoint for key-based authentication. Will be obfuscated during read. Only used if the `authenticationType` is "KeyBased".')
+@secure()
+param secondaryConnectionString string = ''
+
+@description('Optional. The managed identity definition for this resource.')
+param managedIdentities managedIdentitiesType
+
+var identity = !empty(managedIdentities) ? {
```
Comment (AlexanderSehr) [inline avm/res/digital-twins/digital-twins-instance/endpoint--service-bus/main.bicep:42]: Same comment like in `endpoint--event-hub` regarding a potential warning.

File: avm/res/digital-twins/digital-twins-instance/main.bicep [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+
+@description('Optional. Enable/Disable usage telemetry for module.')
+param enableTelemetry bool = true
+
+@description('Optional. Array of role assignment objects that contain the \'roleDefinitionIdOrName\' and \'principalIds\' to define RBAC role assignments on this resource. In the roleDefinitionIdOrName attribute, you can provide either the display name of the role definition, or its fully qualified ID in the following format: \'/providers/Microsoft.Authorization/roleDefinitions/c2f4ef07-c644-48eb-af81-4b1b4947fb11\'.')
+param roleAssignments roleAssignmentType
+
+var formattedUserAssignedIdentities = reduce(map((managedIdentities.?userAssignedResourceIds ?? []), (id) => { '${id}': {} }), {}, (cur, next) => union(cur, next)) // Converts the flat array to an object like { '${id1}': {}, '${id2}': {} }
+
+var identity = !empty(managedIdentities) ? {
```
Comment (AlexanderSehr) [inline avm/res/digital-twins/digital-twins-instance/main.bicep:53]: Same question as in ` endpoint--event-hub` regarding warning

File: avm/res/digital-twins/digital-twins-instance/main.bicep [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+}
+
+resource digitalTwinsInstance_diagnosticSettings 'Microsoft.Insights/diagnosticSettings@2021-05-01-preview' = [for (diagnosticSetting, index) in (diagnosticSettings ?? []): {
+  name: diagnosticSetting.?name ?? '${name}-diagnosticSettings'
+  properties: {
+    storageAccountId: diagnosticSetting.?storageAccountResourceId
+    workspaceId: diagnosticSetting.?workspaceResourceId
+    eventHubAuthorizationRuleId: diagnosticSetting.?eventHubAuthorizationRuleResourceId
+    eventHubName: diagnosticSetting.?eventHubName
+    metrics: diagnosticSetting.?metricCategories ?? [
```
Comment (AlexanderSehr) [inline avm/res/digital-twins/digital-twins-instance/main.bicep:197]: Not implementing the latest schema. ... Sorry ;)

File: avm/res/digital-twins/digital-twins-instance/main.bicep [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+output name string = digitalTwinsInstance.name
+
+@description('The hostname of the Digital Twins Instance.')
+output hostname string = digitalTwinsInstance.properties.hostName
+
+@description('The location the resource was deployed into.')
+output location string = digitalTwinsInstance.location
+
+@description('The principal ID of the system assigned identity.')
+output systemAssignedMIPrincipalId string = (managedIdentities.?systemAssigned ?? false) && contains(digitalTwinsInstance.identity, 'principalId') ? digitalTwinsInstance.identity.principalId : ''
```
Comment (AlexanderSehr) [inline avm/res/digital-twins/digital-twins-instance/main.bicep:246]:
```suggestion
output systemAssignedMIPrincipalId string = digitalTwinsInstance.?identity.?principalId ?? ''
```

File: avm/res/digital-twins/digital-twins-instance/tests/e2e/max/main.test.bicep [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+}
+
+module nestedDependencies 'dependencies.bicep' = {
+  scope: resourceGroup
+  name: '${uniqueString(deployment().name, resourceLocation)}-nestedDependencies'
+  params: {
+    location: resourceLocation
+    virtualNetworkName: 'dep-${namePrefix}-vnet-${serviceShort}'
+    managedIdentityName: 'dep-${namePrefix}-msi-${serviceShort}'
+    eventHubName: 'dt-${uniqueString(serviceShort)}-evh-01'
```
Comment (AlexanderSehr) [inline avm/res/digital-twins/digital-twins-instance/tests/e2e/max/main.test.bicep:41]: Why the uniqueString?

File: avm/res/digital-twins/digital-twins-instance/tests/e2e/max/main.test.bicep [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+      kind: 'CanNotDelete'
+      name: 'myCustomLockName'
+    }
+    privateEndpoints: [
+      {
+        privateDnsZoneResourceIds: [
+          nestedDependencies.outputs.privateDNSZoneResourceId
+        ]
+        subnetResourceId: nestedDependencies.outputs.subnetResourceId
+      }
```
Comment (AlexanderSehr) [inline avm/res/digital-twins/digital-twins-instance/tests/e2e/max/main.test.bicep:123]: Please add a second endpoint that deploys to a second subnet, to ensure vira testing that we don't need to deploy the PE's via `batchSize(1)`

File: avm/res/digital-twins/digital-twins-instance/tests/e2e/max/main.test.bicep [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+    ]
+    roleAssignments: [
+      {
+        roleDefinitionIdOrName: 'Reader'
+        principalId: nestedDependencies.outputs.managedIdentityPrincipalResourceId
+        principalType: 'ServicePrincipal'
+      }
```
Comment (AlexanderSehr) [inline avm/res/digital-twins/digital-twins-instance/tests/e2e/max/main.test.bicep:130]:
```suggestion
      {
        roleDefinitionIdOrName: 'Owner'
        principalId: nestedDependencies.outputs.managedIdentityPrincipalId
        principalType: 'ServicePrincipal'
      }
      {
        roleDefinitionIdOrName: 'b24988ac-6180-42a0-ab88-20f7382dd24c'
        principalId: nestedDependencies.outputs.managedIdentityPrincipalId
        principalType: 'ServicePrincipal'
      }
      {
        roleDefinitionIdOrName: subscriptionResourceId('Microsoft.Authorization/roleDefinitions', 'acdd72a7-3385-48ef-bd42-f606fba81ae7')
        principalId: nestedDependencies.outputs.managedIdentityPrincipalId
        principalType: 'ServicePrincipal'
      }
```

File: avm/res/digital-twins/digital-twins-instance/main.bicep [hunk 9]
```diff
@@ -0,0 +1,399 @@
...
+    privateDnsZoneGroupName: privateEndpoint.?privateDnsZoneGroupName
+    privateDnsZoneResourceIds: privateEndpoint.?privateDnsZoneResourceIds
+    roleAssignments: privateEndpoint.?roleAssignments
+    tags: privateEndpoint.?tags ?? tags
+    customDnsConfigs: privateEndpoint.?customDnsConfigs
+    ipConfigurations: privateEndpoint.?ipConfigurations
+    applicationSecurityGroupResourceIds: privateEndpoint.?applicationSecurityGroupResourceIds
+    customNetworkInterfaceName: privateEndpoint.?customNetworkInterfaceName
+  }
+}]
```
Comment (AlexanderSehr) [inline avm/res/digital-twins/digital-twins-instance/main.bicep:179]:
```suggestion
module digitalTwinsInstance_privateEndpoints 'br/public:avm/res/network/private-endpoint:0.4.0' = [for (privateEndpoint, index) in (privateEndpoints ?? []): {
  name: '${uniqueString(deployment().name, location)}-DigitalTwinsInstance-PrivateEndpoint-${index}'
  scope: resourceGroup(privateEndpoint.?resourceGroupName ?? '')
  params: {
    name: privateEndpoint.?name ?? 'pep-${last(split(digitalTwinsInstance.id, '/'))}-${privateEndpoint.?service ?? 'API'}-${index}'
    privateLinkServiceConnections: privateEndpoint.?manualPrivateLinkServiceConnections != true ? [
      {
        name: privateEndpoint.?privateLinkServiceConnectionName ?? '${last(split(digitalTwinsInstance.id, '/'))}-${privateEndpoint.?service ?? 'API'}-${index}'
        properties: {
          privateLinkServiceId: digitalTwinsInstance.id
          groupIds: [
            privateEndpoint.?service ?? 'API'
          ]
        }
      }
    ] : null
    manualPrivateLinkServiceConnections: privateEndpoint.?manualPrivateLinkServiceConnections == true ? [
      {
        name: privateEndpoint.?privateLinkServiceConnectionName ?? '${last(split(digitalTwinsInstance.id, '/'))}-${privateEndpoint.?service ?? 'API'}-${index}'
        properties: {
          privateLinkServiceId: digitalTwinsInstance.id
          groupIds: [
            privateEndpoint.?service ?? 'API'
          ]
          requestMessage: privateEndpoint.?manualConnectionRequestMessage ?? 'Manual approval required.'
        }
      }
    ] : null
    subnetResourceId: privateEndpoint.subnetResourceId
    enableTelemetry: privateEndpoint.?enableTelemetry ?? enableTelemetry
    location: privateEndpoint.?location ?? reference(split(privateEndpoint.subnetResourceId, '/subnets/')[0], '2020-06-01', 'Full').location
    lock: privateEndpoint.?lock ?? lock
    privateDnsZoneGroupName: privateEndpoint.?privateDnsZoneGroupName
    privateDnsZoneResourceIds: privateEndpoint.?privateDnsZoneResourceIds
    roleAssignments: privateEndpoint.?roleAssignments
    tags: privateEndpoint.?tags ?? tags
    customDnsConfigs: privateEndpoint.?customDnsConfigs
    ipConfigurations: privateEndpoint.?ipConfigurations
    applicationSecurityGroupResourceIds: privateEndpoint.?applicationSecurityGroupResourceIds
    customNetworkInterfaceName: privateEndpoint.?customNetworkInterfaceName
  }
}]
```

File: avm/res/digital-twins/digital-twins-instance/endpoint--event-hub/main.bicep (hunk 1 above)
Comment (ChrisSidebotham) [reply avm/res/digital-twins/digital-twins-instance/endpoint--event-hub/main.bicep:42]: you say implemented but the PR is still open ;)

File: avm/res/digital-twins/digital-twins-instance/main.bicep (hunk 4 above)
Comment (ChrisSidebotham) [reply avm/res/digital-twins/digital-twins-instance/main.bicep:197]: implemted

File: avm/res/digital-twins/digital-twins-instance/README.md [hunk 10]
```diff
@@ -33,7 +33,8 @@ The following section provides usage examples for the module, which were used to
 
 - [Using only defaults](#example-1-using-only-defaults)
 - [Using large parameter set](#example-2-using-large-parameter-set)
-- [WAF-aligned](#example-3-waf-aligned)
+- [Using large parameter set](#example-3-using-large-parameter-set)
```
Comment (AlexanderSehr) [inline avm/res/digital-twins/digital-twins-instance/README.md:36]: Seems like you accidently duplicated the max test?

File: avm/res/digital-twins/digital-twins-instance/tests/e2e/pe/main.test.bicep [hunk 11]
```diff
@@ -0,0 +1,87 @@
+targetScope = 'subscription'
+
+metadata name = 'Using large parameter set'
+metadata description = 'This instance deploys the module with most of its features enabled.'
```
Comment (AlexanderSehr) [inline avm/res/digital-twins/digital-twins-instance/tests/e2e/pe/main.test.bicep:4]: Careful, if you deploy the PE also in a seperate deployment test, please update the metadata to call it out

File: avm/res/digital-twins/digital-twins-instance/tests/e2e/pe/main.test.bicep [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+  name: '${uniqueString(deployment().name, resourceLocation)}-diagnosticDependencies'
+  params: {
+    storageAccountName: 'dep${namePrefix}diasa${serviceShort}03'
+    logAnalyticsWorkspaceName: 'dep-${namePrefix}-law-${serviceShort}'
+    eventHubNamespaceEventHubName: 'dep-${uniqueString(serviceShort)}-evh-01'
+    eventHubNamespaceName: 'dep-${uniqueString(serviceShort)}-evh-01'
+    location: resourceLocation
+  }
+}
+
```
Comment (AlexanderSehr) [inline avm/res/digital-twins/digital-twins-instance/tests/e2e/pe/main.test.bicep:56]:
```suggestion
```

File: avm/res/digital-twins/digital-twins-instance/tests/e2e/pe/main.test.bicep [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+        privateDnsZoneResourceIds: [
+          nestedDependencies.outputs.privateDNSZoneResourceId
+        ]
+        subnetResourceId: nestedDependencies.outputs.pepTestSubnetResourceId
+      }
+    ]
+  }
+  dependsOn: [
+    nestedDependencies
+    diagnosticDependencies
```
Comment (AlexanderSehr) [inline avm/res/digital-twins/digital-twins-instance/tests/e2e/pe/main.test.bicep:85]:
```suggestion
```

File: avm/res/digital-twins/digital-twins-instance/main.bicep (hunk 9 above)
Comment (ChrisSidebotham) [reply avm/res/digital-twins/digital-twins-instance/main.bicep]: Updated in latest push

File: avm/res/digital-twins/digital-twins-instance/endpoint--event-hub/main.bicep (hunk 1 above)
Comment (ChrisSidebotham) [reply avm/res/digital-twins/digital-twins-instance/endpoint--event-hub/main.bicep:44]: Updated in latets push

File: avm/res/digital-twins/digital-twins-instance/endpoint--service-bus/main.bicep (hunk 2 above)
Comment (ChrisSidebotham) [reply avm/res/digital-twins/digital-twins-instance/endpoint--service-bus/main.bicep:44]: updated

File: avm/res/digital-twins/digital-twins-instance/main.bicep (hunk 3 above)
Comment (ChrisSidebotham) [reply avm/res/digital-twins/digital-twins-instance/main.bicep:53]: updated

File: avm/res/digital-twins/digital-twins-instance/tests/e2e/pe/main.test.bicep (hunk 11 above)
Comment (ChrisSidebotham) [reply avm/res/digital-twins/digital-twins-instance/tests/e2e/pe/main.test.bicep:4]: updated

File: avm/res/digital-twins/digital-twins-instance/main.bicep (hunk 3 above)
Comment (ChrisSidebotham) [reply avm/res/digital-twins/digital-twins-instance/main.bicep:53]: some warnings still exist for `identity.` as we discussed previously this is an RP issue

File: avm/res/digital-twins/digital-twins-instance/tests/e2e/pe/main.test.bicep (hunk 13 above)
Comment (ChrisSidebotham) [reply avm/res/digital-twins/digital-twins-instance/tests/e2e/pe/main.test.bicep:85]: done

File: avm/res/digital-twins/digital-twins-instance/main.bicep (hunk 9 above)
Comment (AlexanderSehr) [reply avm/res/digital-twins/digital-twins-instance/main.bicep]: Almost. The scope statement is missing :)

Adding a new comment

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**